Guide: Securely store Micronaut Application secrets in Azure Key Vault (#1099)
micronaut-projects/micronaut-guides: guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc

common:header.adoc[]

common:requirements.adoc[]

common:azure-account-paid.adoc[]

external:micronaut-cloud-database-base/create-app.adoc[]

== Micronaut Azure Key Vault Dependencies

Add the following dependency:

:dependencies:

dependency:micronaut-azure-secret-manager[groupId=io.micronaut.azure]

:dependencies:

== Update the Application configuration to use Azure Key Vault secrets

=== Add the Azure Key Vault secret placeholders to the Application Configuration
[,diff]
.src/main/resources/application.yml
----
micronaut:
  application:
    name: micronautguide

datasources:
  default:
+   username: ${JDBC-USER}
+   password: ${JDBC-PASSWORD}
    dialect: MYSQL

>     driverClassName: ${JDBC_DRIVER:com.mysql.cj.jdbc.Driver}

Reviewer: underscore is used here, is that correct?

Author: Dashes are used because of the Azure naming convention that does not allow underscores. There is a note below this section in the Guide covering this Azure restriction.

`The Azure Key Vault secrets naming convention does not allow "_" to be used as a word joiner. Therefore, when adding JDBC secret placeholders, you should use "-" instead (e.q. JDBC-USER, JDBC-PASSWORD).`

Author: <img width="738" alt="Underscore in Azure secret names" src="https://user-images.githubusercontent.com/4265345/191248299-44a80a0f-2a1d-40c5-b908-9a3dda488a3d.png">

Reviewer: I am aware, but the reference above an underscore is used

Author: Ah, I see your point.

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/6a896fc232dd1a42e605a664cb7488b9ffbbe258.

=== Enable Distributed Configuration

This is accomplished by creating a `bootstrap.yml` file in the `resources` directory and adding the following:

[,yml]
.src/main/resources/bootstrap.yml
----
micronaut:
  application:
    name: micronautguide # <1>
  config-client:
    enabled: true # <2>

azure:
  credential:
    client-secret:
      client-id: ${AZURE_CLIENT_ID} # <3>
      tenant-id: ${AZURE_TENANT_ID} # <4>
      secret: ${AZURE_SECRET_VALUE} # <5>
  key-vault:
    vault-url: ${AZURE_VAULT_URI} # <6>
----
<1> The application name entry is moved from `src/main/resources/application.yml` for resolving distributed configurations properties
<2> Flag to enable the distributed configurations
<3> The client ID of the trusted application that you will register in Azure Portal
<4> The tenant ID
<5> The secret value of the registered application
<6> The Azure Key Vault resource URI

NOTE: More details on how to obtain these values are provided in the Azure Portal section below.

https://docs.micronaut.io/latest/guide/#bootstrap[Click here to get more details on distributed configuration]

> === Clean up of the Application Configuration

Reviewer 2: ```suggestion
=== Clean up the Application Configuration
```

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

Remove the redundant application name entry from the `application.yml`.

[,diff]
.src/main/resources/application.yml
----
- micronaut:
-  application:
-    name: micronautguide

datasources:
  default:
    username: ${JDBC-USER}
    password: ${JDBC-PASSWORD}
    dialect: MYSQL
    driverClassName: ${JDBC-DRIVER:com.mysql.cj.jdbc.Driver}

flyway:
  datasources:
    default:
      enabled: true
----

== Store JDBC secrets in Azure Key Vault using Azure Portal

Create Azure Key Vault service resource.
{empty} +

image::azure-key-vault/azure-vault-create-vault-resouce.png[]
{empty} +

> When Key Vault service deployment is ready, go to the resource in order to create secrets. Prior to clicking on the Secrets option in the left menu list, copy the Vault URI from the main screen section. You will need to set it as an environment variable in order to integrate Azure Key Vault into your application.

Reviewer 2: ```suggestion
When Key Vault service deployment is ready, go to the resource to create secrets. Before clicking on the Secrets option in the left menu list, copy the Vault URI from the main screen section. You will need to set it as an environment variable to integrate Azure Key Vault into your application.
```

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

> When Key Vault service deployment is ready, please go to the resource in order to create secrets. Prior to clicking on the Secrets option in the left menu list, please copy the Vault URI from the main screen section. You will need to set it as an environment variable in order to integrate Azure Key Vault into your application.

Reviewer: you can remove the use of the word "please" in this paragraph

Author: Will do.

Author: Covered by https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/8fed9a79fe5adb0859856581038fe14bef237f02.

{empty} +

image::azure-key-vault/azure-vault-obtain-vault-uri.png[]
{empty} +

Create new secrets for JDBC user and password. Click on the Secrets option in the left menu, then click on Generate/Import.
{empty} +

image::azure-key-vault/azure-vault-create-secret-jdbc-user.png[]
{empty} +

image::azure-key-vault/azure-vault-create-secret-jdbc-password.png[]
{empty} +

> Go to App registration in order to register your application with the Microsoft identity platform. After registering your application,

Reviewer 2: ```suggestion
Go to App registration to register your application with the Microsoft identity platform. After registering your application,
```

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

> copy the secret Application (client) ID, and Directory (tenant) ID. You will use them later on as environment variables in order to connect to Azure Key Vault.

Reviewer 2: ```suggestion
copy the secret Application (client) ID, and Directory (tenant) ID. You will use them later as environment variables to connect to Azure Key Vault.
```

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

{empty} +

image::azure-key-vault/azure-register-application.png[]
{empty} +

> Navigate to your application profile in order to generate application credentials that you will use to authenticate with the Azure Key Vault service. In the left menu, click on the Certificates & secrets. To generate the application credentials, click on the Client secrets tab and then on the New client secret button. Copy the secret value. You will need to set it along the Application (client) ID, Directory (tenant) ID, and Vault URI as environment variables.

Reviewer 2: ```suggestion
Navigate to your application profile to generate application credentials that you will use to authenticate with the Azure Key Vault service. In the left menu, click on the Certificates & secrets. To generate the application credentials, click on the Client secrets tab and then on the New client secret button. Copy the secret value. You must set it along the Application (client) ID, Directory (tenant) ID, and Vault URI as environment variables.
```

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

NOTE: The client's secret value cannot be viewed, except immediately after creation. Be sure to save the secret when created before leaving the page
{empty} +

image::azure-key-vault/azure-register-application-add-client-secret.png[]
{empty} +

> In order to finish the Key Vault setup process. you will need to assign a proper policy to the Key Vault resource. Go to your Key Vault resource profile. Click on the Access policies option in the left menu. Select Get and List options from the Secret Management Operations list. Finally, create an access policy.

Reviewer 2: ```suggestion
To finish the Key Vault setup process. You will need to assign a proper policy to the Key Vault resource. 

* Go to your Key Vault resource profile. 
* Click on the Access policies option in the left menu. 
* Select Get and List options from the Secret Management Operations list. 
* Finally, create an access policy.
```

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

{empty} +

> image::azure-key-vault/azure-vault-create-policy.png[]

Reviewer: the policy should define the minimum amount of access. Is not just "List" and "Get" enough?

Author: Yes, I think you are right. I will make the changes.

Author: Covered by https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/8fed9a79fe5adb0859856581038fe14bef237f02.

{empty} +

Prior to completing the policy creation process, a security principal should be assigned. The policy should refer to your application acting as a security principal.
{empty} +

image::azure-key-vault/azure-vault-create-principal.png[]
{empty} +

== Running the Application

With almost everything in place, you can start the application and try it out. First, set environment variables to configure the application datasource, then start the application.

> Create environment variables for `AZURE_CLIENT_ID`, `AZURE_TENANT_ID`,  `AZURE_SECRET_VALUE`, and `AZURE_VAULT_URI` which will be used in the Micronaut app's `application.yml` datasource:

Reviewer 2: ```suggestion
Create environment variables for `AZURE_CLIENT_ID`, `AZURE_TENANT_ID`,  `AZURE_SECRET_VALUE`, and `AZURE_VAULT_URI`, which will be used in the Micronaut app's `application.yml` datasource:
```

Author: Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.
